pkg/logql/rangemapper.go

```diff
@@ -0,0 +1,332 @@
+			Grouping:  overrideDownstream.Grouping,
+			Operation: overrideDownstream.Operation,
+		}
+	}
+	return &syntax.BinOpExpr{
+		SampleExpr: &syntax.VectorAggregationExpr{
+			Left: m.mapConcatSampleExpr(downstreamExpr, rangeInterval),
+			Grouping: &syntax.Grouping{
+				Without: true,
+			},
+			Operation: syntax.OpTypeSum,
+		},
+		RHS:  &syntax.LiteralExpr{Val: rangeInterval.Seconds()},
+		Op:   syntax.OpTypeDiv,
+		Opts: &syntax.BinOpOptions{},
+	}
+}
+
+// vectorAggrWithRangeDownstreams returns an expression that aggregates a concat sample expression of multiple range
+// aggregations. If a vector aggregation is pushed down, the downstream queries of the concat sample expression are
+// wrapped in the vector aggregation of the parent node.
+// Example:
+// min(bytes_over_time({job="bar"} [2m])
+// => min without (bytes_over_time({job="bar"} [1m]) ++ bytes_over_time({job="bar"} [1m] offset 1m))
+// min by (app) (bytes_over_time({job="bar"} [2m])
+// => min without (min by (app) (bytes_over_time({job="bar"} [1m])) ++ min by (app) (bytes_over_time({job="bar"} [1m] offset 1m)))
+func (m RangeMapper) vectorAggrWithRangeDownstreams(expr *syntax.RangeAggregationExpr, vectorAggrPushdown *syntax.VectorAggregationExpr, op string, rangeInterval time.Duration) syntax.SampleExpr {
+	grouping := expr.Grouping
+	if expr.Grouping == nil {
+		grouping = &syntax.Grouping{
+			Without: true,
+		}
+	}
+	var downstream syntax.SampleExpr = expr
+	if vectorAggrPushdown != nil {
+		downstream = vectorAggrPushdown
+	}
+	return &syntax.VectorAggregationExpr{
+		Left:      m.mapConcatSampleExpr(downstream, rangeInterval),
+		Grouping:  grouping,
+		Operation: op,
+	}
+}
+
+// appendDownstream adds expression expr with a range interval 'interval' and offset 'offset'  to the downstreams list.
+// Returns the updated downstream ConcatSampleExpr.
+func appendDownstream(downstreams *ConcatSampleExpr, expr syntax.SampleExpr, interval time.Duration, offset time.Duration) *ConcatSampleExpr {
+	sampleExpr, _ := clone(expr)
+	sampleExpr.Walk(func(e interface{}) {
+		switch concrete := e.(type) {
+		case *syntax.RangeAggregationExpr:
+			concrete.Left.Interval = interval
+			if offset != 0 {
+				concrete.Left.Offset += offset
+			}
+		}
+	})
+	downstreams = &ConcatSampleExpr{
+		DownstreamSampleExpr: DownstreamSampleExpr{
+			SampleExpr: sampleExpr,
+		},
+		next: downstreams,
+	}
+	return downstreams
+}
+
+// mapConcatSampleExpr transform expr in multiple downstream subexpressions split by offset range interval
+// rangeInterval should be greater than m.splitByInterval, otherwise the resultant expression
+// will have an unnecessary aggregation operation
+func (m RangeMapper) mapConcatSampleExpr(expr syntax.SampleExpr, rangeInterval time.Duration) syntax.SampleExpr {
+	interval := 0
+	splitCount := int(rangeInterval / m.splitByInterval)
+
+	if splitCount == 0 {
+		return expr
+	}
+
+	var downstreams *ConcatSampleExpr
+	for interval = 0; interval < splitCount; interval++ {
```

Comment:
super small nit: `interval` is a bit confusing here since it's an iterator while `rangeInterval` and `m.splitByInterval` are both durations. Maybe call this `split`, since it's an iterator across the splits in the `splitCount`?